012 add role

james-walshe opened the pull request:
- Created UI form for adding a new job role.
- Added front end validation and error messages.
- Wrote unit tests for validator and service classes.

Commit:
Created UI form for adding new job role

Commit:
Added code to populate select menus on form

Commit:
Adding tests for 201 and 500

Commit:
Fixed mockadapter issue

Commit:
Removed console log line

rorkainos commented on .vscode/launch.json:
```diff
@@ -0,0 +1,17 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
I think you should delete this, and add .vscode to your git ignore

rorkainos commented on test/unit/validator/JobRoleValidator.js:
```diff
@@ -0,0 +1,105 @@
+var chai = require('chai');
```
add tests to name of this file

rorkainos commented on app/validator/JobRoleValidator.js:
```diff
@@ -0,0 +1,47 @@
+// provides UI validation for a job role
+module.exports.validateJobRole = function (jobRole) {
+
+    let error = {};
```
instead of returning each time you could continue to get all the errors in one go. Imagine on the ui everything is wrong there user has to change and upload once at a time to get each individual error

rorkainos commented on app/validator/JobRoleValidator.js:
```diff
@@ -0,0 +1,47 @@
+module.exports.validateJobRole = function (jobRole) {
+
+    let error = {};
+
```
perhaps also validate for a blank string

rorkainos commented on app/validator/JobRoleValidator.js:
```diff
@@ -0,0 +1,47 @@
+    // job spec url validation
+    if (jobRole.jobSpecURL.length > 0) {
+        try {
+            // test if url is valid
```
very good, I wouldn't of thought of this

rorkainos commented on app/views/addjobrole.html:
```diff
@@ -0,0 +1,154 @@
+
+  {# SCRIPT FOR COUNTING CHARACTERS IN SPEC TEXT AREA #}
+  <script>
+
```
do this count down the characters like twitter tweet, if so very nice addition

james-walshe replied:
yeah nifty enough

rorkainos commented on test/unit/service/JobServiceTests.js:
```diff
@@ -1,49 +1,95 @@
 var axios = require('axios');
 var MockAdapter = require('axios-mock-adapter');
```
careful with doing git add . as some stuff was pushed that probs shouldn't. you can do git status then git add {file path}

rorkainos commented on test/unit/validator/JobRoleValidator.js:
```diff
@@ -0,0 +1,105 @@
+                jobSpec: "New job in Kainos",
+                jobSpecURL: "https://www.samanthaming.com/tidbits/94-how-to-check-if-object-is-empty/",
+                jobFamily: 1,
+                bandLevel: 2
```
if this was " " it would have passed. I think here you can do something like String.trim(), removes whitespaces either side of the main string ie " test test " becomes "test test"

rorkainos commented on test/unit/validator/JobRoleValidator.js:
```diff
@@ -0,0 +1,105 @@
+        it('should return error when job spec is longer than 1000 characters', () => {
+            let jobRole = {
+                jobRoleName: "Software Engineer",
+                jobSpec: "awawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawawa ...
```
I think here you should just look up some String creation library that allows you to pass a number which will be the length of the string, will look cleaner

Commit:
Made requested changes

Commit:
Made requested changes

shaunganley commented on app/routes.js:
```diff
@@ -9,12 +11,58 @@ router.get('/jobroles', async (req, res) => {
+        req.body.bandLevel = JSON.parse(req.body.bandLevel).bandLevelID;
+        
+        // insert new job
+        js.insertJobRole(req.body)
```
what happens when this fails? should be surrounded with a try/catch

shaunganley commented on app/service/JobService.js:
```diff
@@ -1,14 +1,41 @@
 axios.defaults.baseURL = process.env.API_URL;
 
-module.exports.GET_JOB_ROLES =  '/hr/job-roles';
+module.exports.GET_JOB_ROLES = '/hr/job-roles';
+module.exports.POST_JOB_ROLE = '/hr/job-roles';
```
these will always be the same, just make var generic job roles

shaunganley commented on app/service/JobService.js:
```diff
@@ -1,14 +1,41 @@
+    try {
+        // post request to add new job role
+        const response = await axios.post(this.POST_JOB_ROLE, data);
+        return response;
```
I don't think you will be using the response from this call so no need to return anything

james-walshe replied:
Needed currently to test for a 201 status code. Is there a better way to do this?

shaunganley reviewed (changes requested):
great work, just a few small comments

james-walshe committed:
Create JOB_ROLE_ENDPOINT var

Create JOB_ROLE_ENDPOINT var to replace the GET_JOB_ROLE and POST_JOB_ROLE vars.

Commit:
Made requested changes

Commit:
Merge branch '012-add-role' of github.com:rorkainos/Sprint_One_Front into 012-add-role

james-walshe committed:
Remove console log

james-walshe committed:
Merge branch 'main' into 012-add-role

Commit:
Added bracket that was removed while resolving conflicts

Commit:
Fixed failing test

Commit:
Added sucess message after job role inserted

Commit:
Added element ID's for testing

rorkainos committed:
Merge branch 'main' into 012-add-role
